host/frontend/host-orchestrator/jsonws.go

```diff
@@ -15,44 +15,86 @@
 package main
 
 import (
+	"encoding/json"
 	"log"
+	"net"
 	"net/http"
 	"sync"
 
 	"github.com/gorilla/websocket"
 )
 
-// A websocket connection that can send and receive json objects.
+// A websocket connection that can send and receive JSON objects.
 // Only one thread should call Recv() at a time, Send() and Close() are thread safe
-type JsonWs struct {
+type JSONWs struct {
 	conn     *websocket.Conn
 	writeMtx sync.Mutex
 }
 
 var upgrader = websocket.Upgrader{} // default options
 
-func NewJsonWs(w http.ResponseWriter, r *http.Request) *JsonWs {
+func NewJSONWs(w http.ResponseWriter, r *http.Request) *JSONWs {
 	conn, err := upgrader.Upgrade(w, r, nil)
 	if err != nil {
 		log.Println(err)
 		return nil
 	}
-	return &JsonWs{conn: conn}
+	return &JSONWs{conn: conn}
 }
 
-func (ws *JsonWs) Send(val interface{}) error {
+func (ws *JSONWs) Send(val interface{}) error {
 	ws.writeMtx.Lock()
 	defer ws.writeMtx.Unlock()
 	return ws.conn.WriteJSON(val)
 }
 
-func (ws *JsonWs) Recv(val interface{}) error {
+func (ws *JSONWs) Recv(val interface{}) error {
 	return ws.conn.ReadJSON(val)
 }
 
-func (ws *JsonWs) Close() {
+func (ws *JSONWs) Close() {
 	ws.writeMtx.Lock()
 	defer ws.writeMtx.Unlock()
 	ws.conn.WriteMessage(websocket.CloseMessage, []byte{})
 	ws.conn.Close()
 }
+
+// A Unix socket connection (as returned by Accept) that can send recieve JSON objects.
+// Only one thread should call Recv() at a time, Send and Close are thread safe
+type JSONUnix struct {
+	conn     net.Conn
+	writeMtx sync.Mutex
+	buff     []byte
+}
+
+func NewJSONUnix(c net.Conn) *JSONUnix {
+	return &JSONUnix{
+		conn: c,
+		buff: make([]byte, 10240), // 10K should be enough for each msg
+	}
+}
+
+func (c *JSONUnix) Send(val interface{}) error {
+	s, err := json.Marshal(val)
+	if err != nil {
+		return err
+	}
+	c.writeMtx.Lock()
+	defer c.writeMtx.Unlock()
+	_, err = c.conn.Write(s)
+	return err
+}
+
+func (c *JSONUnix) Recv(val interface{}) error {
+	n, err := c.conn.Read(c.buff)
```

Comment:
The c++ side of this implementation is able to detect when its buffer is too small for the seqpacket messages, is it possible to implement the equivalent here?

Reply:
Something that should help with this is to switch out some of the calls for ones that return more specific types:

 - [`net.ListenUnix`](https://pkg.go.dev/net#ListenUnix) rather than `net.Listen` produces a `UnixListener`
 - [`net.UnixListener.AcceptUnix`](https://pkg.go.dev/net#UnixListener) rather than `Accept` produces a `UnixConn`
 - [`net.UnixConn.ReadMsgUnix`](https://pkg.go.dev/net#UnixConn.ReadMsgUnix) produces more information than the `Read` method
 - If necessary, [`net.UnixConn.SyscallConn`](https://pkg.go.dev/net#UnixConn.SyscallConn) produces a [`syscall.RawConn`](https://pkg.go.dev/syscall#RawConn) which can temporarily expose the file descriptor
 -  The raw [`syscall.Recvmsg`](https://pkg.go.dev/syscall#Recvmsg) is available over file descriptors so it's possible to do `MSG_PEEK` and `MSG_TRUNC`

Reply:
Done